service/src/main/java/bio/terra/workspace/service/workspace/flight/RemovePrivateResourceAccessStep.java

```diff
@@ -0,0 +1,73 @@
+package bio.terra.workspace.service.workspace.flight;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import bio.terra.workspace.service.iam.model.ControlledResourceIamRole;
+import bio.terra.workspace.service.resource.controlled.ControlledResource;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RemovePrivateResourceAccessStep implements Step {
+
+  private final SamService samService;
+  private final AuthenticatedUserRequest userRequest;
+  private final String userToRemove;
+
+  private final Logger logger = LoggerFactory.getLogger(RemovePrivateResourceAccessStep.class);
+
+  public RemovePrivateResourceAccessStep(
+      String userToRemove, SamService samService, AuthenticatedUserRequest userRequest) {
+    this.samService = samService;
+    this.userRequest = userRequest;
+    this.userToRemove = userToRemove;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    FlightMap workingMap = context.getWorkingMap();
+    boolean userStillInWorkspace =
+        Optional.ofNullable(
+                workingMap.get(
+                    ControlledResourceKeys.REMOVED_USER_IS_WORKSPACE_MEMBER, Boolean.class))
+            .orElse(false);
+    // This flight is triggered whenever a user loses any role on a workspace. If they are still
+    // a member of the workspace via a group or another role, we do not need to remove their access
+    // to private resources.
+    if (userStillInWorkspace) {
+      return StepResult.getStepResultSuccess();
+    }
+
+    List<ControlledResource> userResources =
+        workingMap.get(
+            ControlledResourceKeys.REVOCABLE_PRIVATE_RESOURCES, new TypeReference<>() {});
+    for (ControlledResource resource : userResources) {
+      // WSM does not store which roles user have on their private resources, so here we attempt
```

Comment:
I wonder if we should teach WSM to know which role the assigned user has on the private resource. This only applies to the assigned user, right?